### PR TITLE
[try_put_and_wait] Part 8: Add implementation for key_matching join_node

### PR DESCRIPTION
**Pull request by kboyarinov**
### Description 
Extend key_matching join_node with try_put_and_wait API + tests

Fixes # - _issue number(s) if exists_

### Type of change

_Choose one or multiple, leave empty if none of the other choices apply_

_Add a respective label(s) to PR if you have permissions_

- [ ] bug fix - _change that fixes an issue_
- [x] new feature - _change that adds functionality_
- [ ] tests - _change in tests_
- [ ] infrastructure - _change in infrastructure and CI_
- [ ] documentation - _documentation update_

### Tests

- [x] added - _required for new features and some bug fixes_
- [ ] not needed

### Documentation

- [ ] updated in # - _add PR number_
- [x] needs to be updated
- [ ] not needed

### Breaks backward compatibility
- [ ] Yes
- [x] No
- [ ] Unknown

### Notify the following users
_List users with `@` to send notifications_

### CODE REVIEW

File: test/tbb/test_join_node_preview.cpp [hunk 1]
```diff
@@ -227,6 +227,102 @@ void test_try_put_and_wait_reserving() {
...
+        // It is expected that the join_node would push the tuple of three copies of first element in start_work_items
+        // And occupy the concurrency of function. Other tuples would be rejected and taken using push-pull protocol
+        // in order of submission
+        std::size_t check_index = 0;
+
+        for (auto item : start_work_items) {
+            CHECK_MESSAGE(processed_items[check_index++] == item, "Unexpected start_work_items processing");
+        }
+
+        CHECK_MESSAGE(processed_items[check_index++] == wait_message, "Unexpected wait_message processing");
```
Comment (vossmjp) [inline test/tbb/test_join_node_preview.cpp:316]:
```suggestion
        CHECK_MESSAGE(processed_items[check_index++] == wait_message, "Unexpected wait_message processing");
        CHECK_MESSAGE(check_index == processed_items.size(), "Unexpected number of messages");
```

File: test/tbb/test_join_node_preview.cpp [hunk 2]
```diff
@@ -227,6 +227,102 @@ void test_try_put_and_wait_reserving() {
...
+
+        g.wait_for_all();
+
+        for (auto item : new_work_items) {
+            CHECK_MESSAGE(processed_items[check_index++] == item, "Unexpected start_work_items processing");
+        }
+    });
```
Comment (vossmjp) [inline test/tbb/test_join_node_preview.cpp:323]:
```suggestion
    });
   CHECK_MESSAGE(check_index == processed_items.size(), "Unexpected number of messages");
```

**Review by vossmjp — COMMENTED** [review]
A few changes

File: include/oneapi/tbb/detail/_flow_graph_types_impl.h [hunk 3]
```diff
@@ -75,38 +75,45 @@ struct make_sequence < 0, S... > {
 
 //! type mimicking std::pair but with trailing fill to ensure each element of an array
 //* will have the correct alignment
-template<typename T1, typename T2, size_t REM>
-struct type_plus_align {
-    char first[sizeof(T1)];
-    T2 second;
-    char fill1[REM];
+template<class U> struct alignment_of {
```
[the remark this answers is not in the document]
Comment (kboyarinov) [reply include/oneapi/tbb/detail/_flow_graph_types_impl.h:76]: Removed

File: include/oneapi/tbb/detail/_flow_graph_types_impl.h [hunk 4]
```diff
@@ -75,38 +75,45 @@ struct make_sequence < 0, S... > {
...
+
+template <typename T1, typename T2>
+struct alignas(alignof(max_alignment_helper_t<T1, T2>)) aligned_pair {
+    char first[sizeof(T1)];
+    T2 second;
+};
+
+#if __TBB_PREVIEW_FLOW_GRAPH_TRY_PUT_AND_WAIT
+template <typename T1, typename T2, typename T3>
+struct alignas(alignof(max_alignment_helper_t<T1, T2, T3>)) aligned_three {
```
[the remark this answers is not in the document]
Comment (kboyarinov) [reply include/oneapi/tbb/detail/_flow_graph_types_impl.h]: Renamed

File: include/oneapi/tbb/detail/_flow_graph_tagged_buffer_impl.h [hunk 5]
```diff
@@ -30,32 +30,92 @@
...
+        ::new(this->first) value_type(v);
+    }
+
+    void create_element(const value_type& v, const message_metainfo& metainfo) {
+        __TBB_ASSERT(this->third.empty(), nullptr);
+        ::new(this->first) value_type(v);
+        this->third = metainfo;
+
+        for (auto waiter : metainfo.waiters()) {
+            waiter->reserve();
```
[the remark this answers is not in the document]
Comment (kboyarinov) [reply include/oneapi/tbb/detail/_flow_graph_tagged_buffer_impl.h]: There is no special pattern in this regard. I tried to use `1` as an explicit argument everywhere. Changed here and in the corresponding release.

**Review by vossmjp — APPROVED** [review]
Looks good.

**Pull request merged**